src/stdlib_experimental_stats_mean.fypp

```diff
@@ -0,0 +1,293 @@
+
+    integer :: i
+
+    res = 0.0_${k1}$
+    do i = 1, size(x)
+        res = res + x(i)
+    enddo
+    res = res / real(size(x), ${k1}$)
+
+end function mean_1_${k1}$_${k1}$
+#:endfor
+
+#:for i1, k1, t1 in ikti
+module function mean_1_${k1}$_dp(x) result(res)
+    ${t1}$, intent(in) :: x(:)
+    real(dp) :: res
+
+    integer :: i
+
+    res = 0.0_dp
+    do i = 1, size(x)
+        res = res + real(x(i), dp)
+    enddo
+    res = res / real(size(x), dp)
+
+end function mean_1_${k1}$_dp
+#:endfor
+
+
+#:for i1, k1, t1 in iktr
+module function mean_2_all_${k1}$_${k1}$(x) result(res)
+    ${t1}$, intent(in) :: x(:,:)
+    ${t1}$ :: res
+
+    integer :: i, i_
+
+    res = 0.0_${k1}$
+    do i_ = 1, size(x, 2)
+     do i = 1, size(x, 1)
+         res = res + x(i,i_)
+     enddo
+    enddo
+    res = res / real(size(x), ${k1}$)
+
+end function mean_2_all_${k1}$_${k1}$
+#:endfor
+
+#:for i1, k1, t1 in ikti
+module function mean_2_all_${k1}$_dp(x) result(res)
+    ${t1}$, intent(in) :: x(:,:)
+    real(dp) :: res
+
+    integer :: i, i_
+
+    res = 0.0_dp
+    do i_ = 1, size(x, 2)
+     do i = 1, size(x, 1)
+         res = res + real(x(i,i_), dp)
+     enddo
+    enddo
+    res = res / real(size(x), dp)
+
+end function mean_2_all_${k1}$_dp
+#:endfor
+
+#:for i1, k1, t1 in iktr
+module function mean_2_${k1}$_${k1}$(x, dim) result(res)
+    ${t1}$, intent(in) :: x(:,:)
+    integer, intent(in) :: dim
+    ${t1}$ :: res(size(x)/size(x, dim))
+
+    integer :: i, i_
+
+    res = 0.0_${k1}$
+
+    select case(dim)
+     case(1)
+        do i_ = 1, size(x, 2)
+         do i = 1, size(x, 1)
+            res(i_) = res(i_) + x(i, i_)
+         end do
+        end do
+     case(2)
+        do i_ = 1, size(x, 2)
+         do i = 1, size(x, 1)
+            res(i) = res(i) + x(i, i_)
+         end do
+        end do
+     case default
+        call error_stop("ERROR (mean): wrong dimension")
+    end select
+
+    res = res / real(size(x, dim), ${k1}$)
+
+end function mean_2_${k1}$_${k1}$
+#:endfor
+
+#:for i1, k1, t1 in ikti
+module function mean_2_${k1}$_dp(x, dim) result(res)
+    ${t1}$, intent(in) :: x(:,:)
+    integer, intent(in) :: dim
+    real(dp) :: res(size(x)/size(x, dim))
+
+    integer :: i, i_
+
+    res = 0.0_dp
+
+    select case(dim)
+     case(1)
+        do i_ = 1, size(x, 2)
+         do i = 1, size(x, 1)
+            res(i_) = res(i_) + real(x(i, i_), dp)
+         end do
+        end do
+     case(2)
+        do i_ = 1, size(x, 2)
+         do i = 1, size(x, 1)
+            res(i) = res(i) + real(x(i, i_), dp)
+         end do
+        end do
+     case default
+        call error_stop("ERROR (mean): wrong dimension")
+    end select
+
+    res = res / real(size(x, dim), dp)
+
+end function mean_2_${k1}$_dp
+#:endfor
+
+
+#:def ranksuffix(rank)
+#{if rank > 0}#(${":" + ",:" * (rank - 1)}$)#{endif}#
+#:enddef
+
+#:def isuffix(rank)
+#:if rank > 0
+i#{for i in range(2,rank)}#${",i"+"_" * (i-1)}$#{endfor}#
+#:endif
+#:enddef
```

Comment:
How complicated would it be to use more standard variable names such as i,j,k,l,m,n...? Maybe it is worth discussing whether we care about the style of auto-generated code.

I cannot find the place where you encoded the 132 line character limit. The generated code seems to obey this rule:
```Fortran
do i______________ = 1, size(x, 15)
     do i_____________ = 1, size(x, 14)
      do i____________ = 1, size(x, 13)
       do i___________ = 1, size(x, 12)
        do i__________ = 1, size(x, 11)
         do i_________ = 1, size(x, 10)
          do i________ = 1, size(x, 9)
           do i_______ = 1, size(x, 8)
            do i______ = 1, size(x, 7)
             do i_____ = 1, size(x, 6)
              do i____ = 1, size(x, 5)
               do i___ = 1, size(x, 4)
                do i__ = 1, size(x, 3)
                 do i_ = 1, size(x, 2)
                  do i = 1, size(x, 1)
                   res = res + x(i,i_,i__,i___,i____,i_____,i______,i_______,i________,i_________,i__________,i___________,i_______&
                       &_____,i_____________,i______________)
                  end do
                 end do
                end do
               end do
              end do
             end do
            end do
           end do
          end do
         end do
        end do
       end do
      end do
     end do
    end do
```

Reply:
Thanks for your review.

> How complicated would it be to use more standard variable names such as i,j,k,l,m,n...? Maybe it is worth discussing whether we care about the style of auto-generated code.

I am not sure that I would call i,j,k,l,... standard variables. Anyway, I started this implementation by using i,j,k,l,.... But afterwards, I changed my idea because it means that it will reserve 15 characters (in these cases) for loop iterations. Also, it was easier for me to use additional "_" in `fypp` because it involves a simple loop.

Maybe @aradi has some suggestions regarding this issue.

> I cannot find the place where you encoded the 132 line character limit. The generated code seems to obey this rule:

It can be changed in `fypp` by using the argument `-l`. Indeed the default is 132. Should we limit it to 120? Note that in the future I would expect that these files are generated by CMake, and would not be directly available on Github.

Reply:
> I am not sure that I would call i,j,k,l,... standard variables. Anyway, I started this implementation by using i,j,k,l,.... But afterwards, I changed my idea because it means that it will reserve 15 characters (in these cases) for loop iterations. Also, it was easier for me to use additional "_" in `fypp` because it involves a simple loop.

I understand your reasoning. It is also easier to understand for the maintainers of the template file. I was only wondering whether we care about "style" (including variable name choice) in the auto-generated files.

Reply:
> > I cannot find the place where you encoded the 132 line character limit. The generated code seems to obey this rule:
> 
> It can be changed in `fypp` by using the argument `-l`. Indeed the default is 132. Should we limit it to 120? Note that in the future I would expect that these files are generated by CMake, and would not be directly available on Github.

In the [style guide](https://github.com/fortran-lang/stdlib/blob/master/STYLE_GUIDE.md) of `stdlib`, it is mentioned that:

>Line length should be limited to 80 characters and must not exceed 132. 

So we could limit `fypp` to 80 characters, or use the default lenght which is 132.

Reply:
@ivan-pi Following @aradi 's [suggestion](https://github.com/fortran-lang/stdlib/pull/119#issuecomment-577829560) I kept the default lenght (132) for the auto-generated files.
I also replaced all the underscores by numbers. This is now much easier to read. Thank you for mentioning this issue!